mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,31 +1,473 @@
-use tracing::info;
+use std::{collections::HashMap, marker::PhantomData, str::FromStr, time::Duration};
 
-use crate::settings::ConnectionSettings;
+use async_trait::async_trait;
+use mqtt3::{proto::Publication, Event, ReceivedPublication};
+use mqtt_broker::TopicFilter;
+use tracing::{debug, info, warn};
+
+use crate::client::{ClientConnectError, EventHandler, MqttClient};
+use crate::persist::{memory::InMemoryPersist, Persist};
+use crate::settings::{ConnectionSettings, Credentials, Topic};
+
+const BATCH_SIZE: usize = 10;
 
 /// Bridge implementation that connects to local broker and remote broker and handles messages flow
 pub struct Bridge {
+    system_address: String,
+    device_id: String,
     connection_settings: ConnectionSettings,
+    forwards: HashMap<String, Topic>,
+    subscriptions: HashMap<String, Topic>,
 }
 
 impl Bridge {
-    pub fn new(connection_settings: ConnectionSettings) -> Self {
+    pub fn new(
+        system_address: String,
+        device_id: String,
+        connection_settings: ConnectionSettings,
+    ) -> Self {
+        let forwards = connection_settings
+            .forwards()
+            .iter()
+            .map(|sub| Self::format_key_value(sub))
+            .collect();
+
+        let subscriptions = connection_settings
+            .subscriptions()
+            .iter()
+            .map(|sub| Self::format_key_value(sub))
+            .collect();
+
         Bridge {
+            system_address,
+            device_id,
             connection_settings,
+            forwards,
+            subscriptions,
         }
     }
 
-    pub async fn start(&self) {
+    fn format_key_value(topic: &Topic) -> (String, Topic) {
+        let key = if let Some(local) = topic.local() {
+            format!("{}/{}", local, topic.pattern().to_string())
+        } else {
+            topic.pattern().into()
+        };
+        (key, topic.clone())
+    }
+
+    pub async fn start(&self) -> Result<(), BridgeError> {
         info!("Starting bridge...{}", self.connection_settings.name());
 
-        self.connect_to_local().await;
-        self.connect_to_remote().await;
+        self.connect_to_local().await?;
+        self.connect_to_remote().await?;
+
+        Ok(())
+    }
+
+    async fn connect_to_remote(&self) -> Result<(), BridgeError> {
+        info!(
+            "connecting to remote broker {}",
+            self.connection_settings.address()
+        );
+
+        self.connect(
+            &self.subscriptions,
```

Comment:
you borrow subsctiptions here but inside connect method you clone everything else. I wonder if it worthwhile to clone hashmap here and just deconstruct it inside connect method. It'll make code easier I believe